API/get all comments

timotiusnc opened the pull request:
Closes #20 

Comments will now be retrieved recursively.
`News.getTopLevelComments()` will return `Observable<List<Comment>>` that contains only top level comments of that News instance, each of it will already contain child comments.

timotiusnc committed:
1. Add function to get all comments recursively
2. Update constructor for Story, Job, and Poll
3. Update RestAdapter to use one ItemTypeAdapter (News only)

timotiusnc committed:
Update Comment's getText to use Math.min

timotiusnc committed:
Set comments in nested data structure

timotiusnc committed:
Merge branch 'master' into API/getAllComments

timotiusnc committed:
Fix NullPointerException when there's deleted top level comments

timotiusnc committed:
Log the comments for review purpose

timotiusnc commented on app/src/main/java/io/pumpkinz/pumpkinreader/data/NewsDetailAdapter.java:
```diff
@@ -75,12 +75,12 @@ public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int i) {
                 break;
             case 1:
-                Comment comment = this.dataset.getComments().get(i - 1);
+                /*Comment comment = this.dataset.getTopLevelComments().get(i - 1);
```
I commented this out because it made the application to be forced close.

wiseodd replied:
I guess it's because getTopLevelComments returns an Observable

timotiusnc replied:
I mean, the UI is not yet ready, right?

wiseodd replied:
It's ready for top level comments

timotiusnc commented on app/src/main/java/io/pumpkinz/pumpkinreader/service/CommentSource.java:
```diff
@@ -0,0 +1,124 @@
+    private CommentSource() {
+    }
+
+    public static CommentSource service() {
```
I'm wondering if there's better pattern than using singleton. I'm starting to think I'm over-using singleton, here.

wiseodd replied:
I might not be the best person to be asked about design pattern and such

wiseodd commented on app/src/main/java/io/pumpkinz/pumpkinreader/model/Comment.java:
```diff
@@ -13,13 +15,17 @@
-    public Comment() {}
+    public Comment() {
+        this.kids = new ArrayList<>();
+        this.childComments = new ArrayList<>();
```
What's the difference between kids and childComments?

timotiusnc replied:
oh well, the naming from HN again.
`kids` is commentIds.
`childComments` is the actual `Comment` object.

wiseodd replied:
Is it the rule of Retrofit to match the field name?
If it's not, better be changed.

timotiusnc replied:
Yes, if not, we have to define our own JSON parser.
What if we change the getter instead? (`getKids` into `getCommentIds`)

wiseodd replied:
Good idea!
After all, the point of encapsulation is to not worry about the internal implementation.

wiseodd commented on app/src/main/java/io/pumpkinz/pumpkinreader/data/NewsDetailAdapter.java:
```diff
@@ -75,12 +75,12 @@ public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int i) {
                 commentViewHolder.getSubmitter().setText(comment.getBy());
                 commentViewHolder.getDate().setText(this.dateFormatter.timeAgo(comment.getTime()));
-                commentViewHolder.getBody().setText(comment.getText());
+                commentViewHolder.getBody().setText(comment.getText());*/
```
Use Html.fromHtml(), because item's body is HTML formatted

timotiusnc committed:
1. Wire DataSource.getComments to NewsDetailFragment
2. Remove CommentSource
3. Update getKids getter into getCommentIds for readability

timotiusnc commented:
Updated!

Please check it, I don't know why the last top level comment is rendered weirdly.

wiseodd commented:
I will work on it while working on hierarchical comment view.
This one is LGTM.
